pkg/build/generator/generator_test.go

```diff
@@ -122,6 +122,218 @@ func TestInstantiateGenerateBuildError(t *testing.T) {
+				Triggers: tc.triggers,
+			},
+		}
+		generator := mockBuildGenerator()
+		client := generator.Client.(Client)
+		client.GetBuildConfigFunc =
+			func(ctx kapi.Context, name string) (*buildapi.BuildConfig, error) {
+				return bc, nil
+			}
+		client.UpdateBuildConfigFunc =
+			func(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error {
+				bc = buildConfig
+				return nil
+			}
+		generator.Client = client
+
+		req := &buildapi.BuildRequest{
+			TriggeredByImage: &kapi.ObjectReference{
+				Kind: "DockerImage",
+				Name: imageID,
+			},
+			From: tc.reqFrom,
+		}
+		_, err := generator.Instantiate(kapi.NewDefaultContext(), req)
+		if err != nil && !tc.errorExpected {
+			t.Errorf("%s: unexpected error %v", tc.name, err)
+			continue
+		}
+		if err == nil && tc.errorExpected {
+			t.Errorf("%s: expected error but didn't get one", tc.name)
+			continue
+		}
+		if tc.errorExpected {
+			continue
+		}
+		for i := range bc.Spec.Triggers {
+			if i == tc.triggerIndex {
+				// Verify that the trigger got updated
+				if bc.Spec.Triggers[i].ImageChange.LastTriggeredImageID != imageID {
+					t.Errorf("%s: expeccted trigger at index %d to contain imageID %s", tc.name, i, imageID)
+				}
+				continue
+			}
+			if bc.Spec.Triggers[i].Type == buildapi.ImageChangeBuildTriggerType {
+				if bc.Spec.Triggers[i].ImageChange.LastTriggeredImageID != "" {
+					t.Errorf("%s: expected LastTriggeredImageID for trigger at %d to be empty", tc.name, i)
+				}
+			}
+		}
+	}
+}
+
+func TestFindImageTrigger(t *testing.T) {
+	defaultTrigger := &buildapi.ImageChangeTrigger{}
+	image1Trigger := &buildapi.ImageChangeTrigger{
+		From: &kapi.ObjectReference{
+			Name: "image1:tag1",
+		},
+	}
+	image2Trigger := &buildapi.ImageChangeTrigger{
+		From: &kapi.ObjectReference{
+			Name:      "image2:tag2",
+			Namespace: "image2ns",
+		},
+	}
+	bc := &buildapi.BuildConfig{
+		Spec: buildapi.BuildConfigSpec{
+			Triggers: []buildapi.BuildTriggerPolicy{
+				{
+					Type: buildapi.GenericWebHookBuildTriggerType,
+				},
+				{
+					Type:        buildapi.ImageChangeBuildTriggerType,
+					ImageChange: defaultTrigger,
+				},
+				{
+					Type:        buildapi.ImageChangeBuildTriggerType,
+					ImageChange: image1Trigger,
+				},
+				{
+					Type:        buildapi.ImageChangeBuildTriggerType,
+					ImageChange: image2Trigger,
+				},
+			},
+		},
+	}
+
+	tests := []struct {
+		input  *kapi.ObjectReference
+		expect *buildapi.ImageChangeTrigger
+	}{
+		{
+			input:  nil,
+			expect: nil,
+		},
+		{
+			input: &kapi.ObjectReference{
+				Name: "image1:tag1",
+			},
+			expect: image1Trigger,
+		},
+		{
+			input: &kapi.ObjectReference{
+				Name:      "image1:tag1",
+				Namespace: "otherns",
+			},
+			expect: defaultTrigger,
+		},
+		{
+			input: &kapi.ObjectReference{
+				Name:      "image2:tag2",
+				Namespace: "image2ns",
+			},
+			expect: image2Trigger,
+		},
+		{
+			input: &kapi.ObjectReference{
+				Name: "image3:tag3",
+			},
+			expect: defaultTrigger,
+		},
```

Comment:
I thought I covered both cases:
1) implied default namespace of the Strategy.From matches the explicit namespace of the trigger.From:
https://github.com/openshift/origin/pull/3748/files#diff-792eff2c126cc6918c8549d221273179R368
2) implied default namespace of trigger.From matches the explicit namespace of the Strategy.From:
https://github.com/openshift/origin/pull/3748/files#diff-792eff2c126cc6918c8549d221273179R361